File: LinkedList/python/problem_109_convert_sorted_list_to_bst.py
"""Problem 109: Convert Sorted List to Binary Search Tree - Medium
Convert sorted linked list to height-balanced BST
Example: [-10,-3,0,5,9] → Height-balanced BST
Time: O(n log n), Space: O(log n)
"""
class ListNode:
    def __init__(self, val=0, next=None):
        self.val, self.next = val, next

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def sortedListToBST(self, head: ListNode) -> TreeNode:
        if not head:
            return None
        if not head.next:
            return TreeNode(head.val)
        
        # Find middle using slow/fast pointers
        slow, fast, prev = head, head, None
        while fast and fast.next:
            prev = slow
            slow = slow.next
            fast = fast.next.next
        
        # Disconnect left half
        if prev:
            prev.next = None
        
        # Middle becomes root
        root = TreeNode(slow.val)
        root.left = self.sortedListToBST(head)
        root.right = self.sortedListToBST(slow.next)
        
        return root

File: LinkedList/java/Problem109_ConvertSortedListToBST.java
/**
 * Problem 109: Convert Sorted List to Binary Search Tree
 * Difficulty: Medium
 * 
 * Given the head of a singly linked list where elements are sorted in ascending order,
 * convert it to a height balanced BST.
 * 
 * Example: [-10,-3,0,5,9] → Height-balanced BST
 * Time: O(n log n), Space: O(log n)
 */
class ListNode {
    int val;
    ListNode next;
    ListNode(int val) { this.val = val; }
}

class TreeNode {
    int val;
    TreeNode left, right;
    TreeNode(int val) { this.val = val; }
}

public class Problem109_ConvertSortedListToBST {
    public TreeNode sortedListToBST(ListNode head) {
        if (head == null) return null;
        if (head.next == null) return new TreeNode(head.val);
        
        // Find middle using slow/fast pointers
        ListNode slow = head, fast = head, prev = null;
        while (fast != null && fast.next != null) {
            prev = slow;
            slow = slow.next;
            fast = fast.next.next;
        }
        
        // Disconnect left half
        if (prev != null) prev.next = null;
        
        // Middle becomes root
        TreeNode root = new TreeNode(slow.val);
        root.left = sortedListToBST(head);
        root.right = sortedListToBST(slow.next);
        
        return root;
    }
}
